File: src/sched_nice.c
#include <sys/types.h>
#include <sys/wait.h>
#include <time.h>
#include <unistd.h>
#include <stdio.h>
#include <stdlib.h>
// unused
// #include <string.h>
#include <err.h>
// need in macos
#include <signal.h>

#define NLOOP_FOR_ESTIMATION 1000000000UL
#define NSECS_PER_MSEC 1000000UL
#define NSECS_PER_SEC 1000000000UL

static inline void load(unsigned long nloop)
{
  unsigned long i;
  for (i = 0; i < nloop; i++)
    ;
}

static inline long diff_nsec(struct timespec before, struct timespec after)
{
  return ((after.tv_sec * NSECS_PER_SEC + after.tv_nsec) - (before.tv_sec * NSECS_PER_SEC + before.tv_nsec));
}

static void child_fn(int id, struct timespec *buf, int nrecord, unsigned long nloop_per_resol, struct timespec start)
{
  int i;
  for (i = 0; i < nrecord; i++)
  {
    struct timespec ts;
    load(nloop_per_resol);
    clock_gettime(CLOCK_MONOTONIC, &ts);
    buf[i] = ts;
  }
  for (i = 0; i < nrecord; i++)
  {
    printf("%d\t%ld\t%d\n", id, diff_nsec(start, buf[i]) / NSECS_PER_MSEC, (i + 1) * 100 / nrecord);
  }
  exit(EXIT_SUCCESS);
}

static unsigned long loops_per_msec()
{
  struct timespec before, after;
  // @see https://qiita.com/ozaki-r/items/fb4a48c2833e4b479ae1
  clock_gettime(CLOCK_MONOTONIC, &before);

  unsigned long i;
  for (i = 0; i < NLOOP_FOR_ESTIMATION; i++)
    ;
  clock_gettime(CLOCK_MONOTONIC, &after);

  int ret;
  return NLOOP_FOR_ESTIMATION * NSECS_PER_MSEC / diff_nsec(before, after);
}

// helper function to validate args
static inline void arg_validation(int arg, char *argname)
{
  if (arg < 1)
  {
    fprintf(stderr, "<%s>(%d) should be >= 1\n", argname, arg);
    exit(EXIT_FAILURE);
  }
}

static pid_t *pids;

int main(int argc, char *argv[])
{
  int ret = EXIT_FAILURE;

  if (argc < 3)
  {
    fprintf(stderr, "usage: %s <total[ms]> <resolution[ms]>\n", argv[0]);
    exit(EXIT_FAILURE);
  }

  int nproc = 2;
  int total = atoi(argv[1]);
  int resol = atoi(argv[2]);

  arg_validation(total, "total");
  arg_validation(resol, "resol");

  if (total % resol)
  {
    fprintf(stderr, "<total>(%d) should be multiple of <resolution>(%d)\n", total, resol);
    exit(EXIT_FAILURE);
  }
  int nrecord = total / resol;

  struct timespec *logbuf = malloc(nrecord * sizeof(struct timespec));
  if (!logbuf)
  {
    err(EXIT_FAILURE, "malloc(logbuf) failed");
  }

  // puts("estimating workload which takes just one millisecond");

  unsigned long nloop_per_resol = loops_per_msec() * resol;

  // puts("end estimation");
  fflush(stdout);

  pids = malloc(nproc * sizeof(pid_t));
  if (pids == NULL)
  {
    warn("malloc(pids) failed");
    goto free_logbuf;
  }

  struct timespec start;
  clock_gettime(CLOCK_MONOTONIC, &start);

  int i, ncreated;
  for (i = 0, ncreated = 0; i < nproc; i++, ncreated++)
  {
    pids[i] = fork();
    if (pids[i] < 0)
    {
      goto wait_children;
    }
    else if (pids[i] == 0)
    {
      // children
      if (i == 1)
      {
        nice(5);
      }
      child_fn(i, logbuf, nrecord, nloop_per_resol, start);
      /* shouldn't reach here */
    }
  }
  ret = EXIT_SUCCESS;

// parent
wait_children:
  if (ret == EXIT_FAILURE)
  {
    for (i = 0; i < ncreated; i++)
    {
      if (kill(pids[i], SIGINT) < 0)
      {
        warn("kill(%d) failed", pids[i]);
      }
    }
  }

  for (i = 0; i < ncreated; i++)
  {
    if (wait(NULL) < 0)
    {
      warn("wait() failed.");
    }
  }

free_logbuf:
  free(logbuf);
  exit(ret);
}

File: rust/src/sched_nice.rs
use nix::{
    libc::{nice, EXIT_FAILURE, EXIT_SUCCESS},
    sys::{
        signal::{kill, Signal::SIGINT},
        time::TimeSpec,
        wait::wait,
    },
    time::{clock_gettime, ClockId},
    unistd::{ForkResult, Pid},
};
use std::{env, process::exit};

const NLOOP_FOR_ESTIMATION: usize = 1_000_000_000;
const NSECS_PER_MSEC: usize = 1_000_000;
const NSECS_PER_SEC: usize = 1_000_000_000;

fn load(nloop: usize) {
    for _ in 0..nloop {}
}

fn child_fn(
    id: usize,
    buf: &mut Vec<TimeSpec>,
    nrecord: usize,
    nloop_per_resol: usize,
    start: TimeSpec,
) {
    for _ in 0..nrecord {
        load(nloop_per_resol);
        buf.push(get_time());
    }
    for i in 0..nrecord {
        println!(
            "{}\t{}\t{}",
            id,
            diff_nsec(start, buf[i]) / NSECS_PER_MSEC,
            (i + 1) * 100 / &nrecord
        );
    }
    std::process::exit(EXIT_SUCCESS);
}

fn get_time() -> TimeSpec {
    match clock_gettime(ClockId::CLOCK_MONOTONIC) {
        Ok(time) => time,
        Err(e) => {
            eprintln!("clock_gettime() failed {:?}", e);
            std::process::exit(EXIT_FAILURE);
        }
    }
}

// 2つの TimeSpec の差分を計算しナノ秒で返却します
fn diff_nsec(before: TimeSpec, after: TimeSpec) -> usize {
    (after.tv_sec() as usize * NSECS_PER_SEC + after.tv_nsec() as usize)
        - (before.tv_sec() as usize * NSECS_PER_SEC + before.tv_nsec() as usize)
}

/// CPU時間を 1ms 使う処理に必要なループ回数を推定します
fn loops_per_msec() -> usize {
    let before = get_time();

    // リリースビルドしてしまうと最適化がかかり機能しなくなる
    for _ in 0..NLOOP_FOR_ESTIMATION {}

    let after = get_time();

    // ループ回数をかかった時間(diff_nsec) でわり、1nsあたりのループ回数を計算し、NSECS_PER_MSECを掛け、単位をmsにする
    NLOOP_FOR_ESTIMATION * NSECS_PER_MSEC / diff_nsec(before, after)
}

/// 与えられた引数を `usize` に変換します。
/// 変換に失敗した場合は、エラーを表示し終了します。
fn arg_validation(arg: &String, argname: &str) -> usize {
    match arg.parse::<usize>() {
        Ok(r) => {
            if r < 1 {
                eprintln!("<{}>({}) should be >= 1", argname, arg);
                std::process::exit(EXIT_FAILURE);
            }
            r
        }
        Err(e) => {
            eprintln!("<{}>({}) should be number: {:?}", argname, arg, e);
            std::process::exit(EXIT_FAILURE);
        }
    }
}

/// ## コマンドライン引数
/// - 第1引数（total）: プログラムを動作させる合計時間（ms単位）
/// - 第2引数（resol）: 統計情報の採取間隔（ms単位）
fn main() {
    let ret;
    let args: Vec<String> = env::args().collect();

    if args.len() < 3 {
        eprintln!("usage: {} <total[ms]> <resolution[ms]>", args[0]);
        exit(EXIT_FAILURE);
    }

    // 並行プロセス数
    let nproc = 2;
    // プログラムを動作させる合計時間(ms)
    let total = arg_validation(&args[1], "total");
    // 統計情報の採取時間(ms)
    let resol = arg_validation(&args[2], "resol");

    // プログラムを動作させる合計時間が統計情報の採取時間(resol: 解像度)で割り切れるかを確認
    if total % resol != 0 {
        eprintln!(
            "<total>({}) should be multiple of <resolution>({})",
            total, resol
        );
    }
    // 計測するレコード数を計算する
    // 100ms を 10ms 単位で計測する場合は、10 レコードとなる
    let nrecord = total / resol;

    // 1ms にかかるループ回数を計測し、それを解像度(ms)に合わせる
    let nloop_per_resol = loops_per_msec() * resol;

    let mut logbuf: Vec<TimeSpec> = Vec::<TimeSpec>::with_capacity(nrecord);
    let mut pids = Vec::<Pid>::with_capacity(nproc);

    let start = get_time();

    let mut ncreated = 0;
    for i in 0..nproc {
        match unsafe { nix::unistd::fork() } {
            Ok(ForkResult::Parent { child, .. }) => {
                pids.push(child);
            }
            Ok(ForkResult::Child) => {
                // 最初に作成したプロセスだけ nice を変更
                if ncreated == 0 {
                    unsafe { nice(5) };
                }
                child_fn(i, &mut logbuf, nrecord, nloop_per_resol, start);
            }
            Err(_) => eprintln!("fork() failed."),
        }
        ncreated += 1;
    }

    ret = EXIT_SUCCESS;

    if ret == EXIT_FAILURE {
        for i in 0..ncreated {
            if let Err(e) = kill(pids[i], SIGINT) {
                eprintln!("kill({}) failed: {:?}", &pids[i], e);
            }
        }
    }

    for _ in 0..ncreated {
        if let Err(e) = wait() {
            eprintln!("wait() failed: {:?}", e);
        }
    }

    std::process::exit(ret);
}
